File: src/ai/conversation/conversation.service.ts
import { LLMService } from '../shared/llm.service'
import { SessionService } from '../memory/session.service'
import { buildConversationSystemPrompt } from './conversation.prompt'
import { AgentSession } from '../shared/agent.types'
import { ExpertiseLevel } from '../shared/constants'

export class ConversationService {
  private llm = new LLMService()
  private sessionService = new SessionService()

  async handle(session: AgentSession, userMessage: string): Promise<string> {

    // 1. Detect expertise level using LLM if not already set
    if (!session.expertiseLevel) {
      session.expertiseLevel = await this.detectExpertiseLevelLLM(userMessage)
      console.log(`🧠 LLM Detected expertise: ${session.expertiseLevel}`)
    }

    // 2. Count conversation rounds
    const conversationRound = Math.floor(session.history.length / 2)

    // 3. Build system prompt
    const systemPrompt = buildConversationSystemPrompt(
      session.expertiseLevel,
      session.project || {},
      session.clientName || 'Client',
      conversationRound
    )

    // 4. Build messages
    const messages = [
      { role: 'system' as const, content: systemPrompt },
      ...session.history,
      { role: 'user' as const, content: userMessage },
    ]

    // 5. Call LLM
    const reply = await this.llm.call(messages)

    // 6. Save to history
    session.history.push({ role: 'user', content: userMessage })
    session.history.push({ role: 'assistant', content: reply })
    await this.sessionService.save(session)

    return reply
  }

  // ── LLM-based expertise detection ────────────────────────
  private async detectExpertiseLevelLLM(message: string): Promise<ExpertiseLevel> {
    try {
      const prompt = `
You are analyzing a client's first message to a freelance platform AI assistant.

CLIENT MESSAGE: "${message}"

Classify the client's technical expertise level based on their message:
- BEGINNER: Vague idea, no technical terms, doesn't know what they need exactly
- INTERMEDIATE: Knows some features they want, mentions some tech or platforms
- ADVANCED: Uses technical terms, knows exact stack, mentions architecture/APIs

RETURN ONLY ONE WORD — exactly one of: BEGINNER, INTERMEDIATE, ADVANCED
No explanation, no punctuation, just the word.`

      const result = await this.llm.call([{ role: 'user', content: prompt }])
      const cleaned = result.trim().toUpperCase()

      if (cleaned.includes('ADVANCED')) return ExpertiseLevel.ADVANCED
      if (cleaned.includes('INTERMEDIATE')) return ExpertiseLevel.INTERMEDIATE
      return ExpertiseLevel.BEGINNER

    } catch (error) {
      console.error('❌ LLM expertise detection failed, using INTERMEDIATE as default')
      return ExpertiseLevel.INTERMEDIATE
    }
  }

  isProjectComplete(project: Partial<any>): boolean {
    return !!(
      project.projectType &&
      project.features?.length > 0 &&
      project.budgetMin &&
      project.budgetMax &&
      project.timeline
    )
  }
}

File: src/ai/conversation/conversation_service.py
from typing import Dict, Any, List
from conversation.conversation_prompt import build_conversation_system_prompt
from shared.constants import (
    ExpertiseLevel,
    UserType,
    UrgencyLevel,
    BudgetSensitivity,
    CommunicationStyle
)
import json
import re


class ConversationService:
    def __init__(self, llm_service, session_service):
        self.llm = llm_service
        self.session_service = session_service

    async def handle(self, session: Dict[str, Any], user_message: str) -> str:

        # 1. Detect user persona
        if not session.get("persona"):
            session["persona"] = await self.detect_user_persona_llm(user_message)
            print(f"👤 Persona detected: {session['persona'].get('userType', 'unknown')}")

        # 2. Conversation round
        history = session.get("history", [])
        conversation_round = len(history) // 2

        # 3. Build prompt
        system_prompt = build_conversation_system_prompt(
            session["persona"],
            session.get("project", {}),
            session.get("clientName", "Client"),
            conversation_round
        )

        # 4. Messages
        messages = [{"role": "system", "content": system_prompt}]
        messages.extend(history)
        messages.append({"role": "user", "content": user_message})

        # 5. LLM call
        reply = await self.llm.call(messages)

        # 6. Save history
        history.append({"role": "user", "content": user_message})
        history.append({"role": "assistant", "content": reply})
        session["history"] = history

        await self.session_service.save(session)

        return reply

    async def detect_user_persona_llm(self, message: str) -> Dict[str, Any]:
        try:
            prompt = f"""
Analyze this user message and determine their persona for a freelance platform.
MESSAGE: "{message}"

PERSONA CATEGORIES:
- expertiseLevel: BEGINNER, INTERMEDIATE, ADVANCED
- userType: business_owner, student, startup, technical, experienced_client, confused
- urgency: low, medium, high
- budgetSensitivity: low, medium, high
- communicationStyle: formal, casual
- primaryGoal: A short string (3-5 words)

RETURN STRICT JSON ONLY:
{{
  "expertiseLevel": "...",
  "userType": "...",
  "urgency": "...",
  "budgetSensitivity": "...",
  "communicationStyle": "...",
  "primaryGoal": "..."
}}
"""
            result = await self.llm.call([{"role": "user", "content": prompt}])
            cleaned = re.sub(r"```json|```", "", result, flags=re.IGNORECASE).strip()
            persona = json.loads(cleaned)
            return persona

        except Exception as e:
            print(f"❌ Persona detection failed: {e}")
            return {
                "expertiseLevel": ExpertiseLevel.INTERMEDIATE,
                "userType": UserType.CONFUSED,
                "urgency": UrgencyLevel.LOW,
                "budgetSensitivity": BudgetSensitivity.MEDIUM,
                "communicationStyle": CommunicationStyle.FORMAL,
                "primaryGoal": "Establish project requirements"
            }

    def is_project_complete(self, project: Dict[str, Any]) -> bool:
        return all([
            project.get("projectType"),
            project.get("features"),
            project.get("budgetMin"),
            project.get("budgetMax"),
            project.get("timeline")
        ])